service/history/historyEngine.go

```diff
@@ -1289,479 +1289,486 @@ Update_History_Loop:
 		sizeChecker.completedID = completedID
 		sizeChecker.msBuilder = msBuilder
 
-	Process_Decision_Loop:
-		for _, d := range request.Decisions {
-			switch *d.DecisionType {
-			case workflow.DecisionTypeScheduleActivityTask:
-				e.metricsClient.IncCounter(metrics.HistoryRespondDecisionTaskCompletedScope,
-					metrics.DecisionTypeScheduleActivityCounter)
-				targetDomainID := domainID
-				attributes := d.ScheduleActivityTaskDecisionAttributes
-				// First check if we need to use a different target domain to schedule activity
-				if attributes.Domain != nil {
-					// TODO: Error handling for ActivitySchedule failed when domain lookup fails
-					domainEntry, err := e.shard.GetDomainCache().GetDomain(*attributes.Domain)
-					if err != nil {
-						return nil, &workflow.InternalServiceError{Message: "Unable to schedule activity across domain."}
+		binChecksum := request.GetBinaryChecksum()
+		if _, ok := domainEntry.GetConfig().BadBinaries.Binaries[binChecksum]; !ok {
```

Comment:
Is there a way to structure this code without having to nest the whole decision loop? This makes the code in this function even more difficult to understand.

Reply:
Another option is to use goto. But what we actually need is to refactor this big for-loop.

Reply:
Make sure you selected "Hide whitespace changes" in the diff setting

Reply:
issue created: https://github.com/uber/cadence/issues/1783

Reply:
maybe switch the if and else.. in the current code it is very difficult to figure out what's the effect of bad binaries existing as its buried after the full process_decision_loop..

Reply:
yes, agreed. Bowei and I had the same discussion. Already updated with that.